### PR TITLE
Batch transactions when notified of committed new entries in raft

### PR DESCRIPTION
**Pull request by davidegrohmann**

### CODE REVIEW

**Comment by lutovich** [comment]
I think nested callbacks in kernel in `TransactionToApply` and `Commitment` look really complex. There is also a hierarchy of callbacks higher in the stack: `ReplicatedTransactionStateMachine#applyCommand()` gets a callback and registers another callback on `TransactionToApply` that calls the first callback.

Let's spend some time to design this better, please.
Adding 'kernel' label for the team to review.

**Comment by davidegrohmann** [comment]
@lutovich I am afraid you cannot really do much more than that since the only thing I have to commit transaction is `TransactionCommitProcess`. Anyhow waiting for your feedback.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/id/ReplicatedIdAllocationStateMachine.java [hunk 1]
```diff
@@ -43,24 +44,22 @@ public ReplicatedIdAllocationStateMachine( StateStorage<IdAllocationState> stora
     }
 
     @Override
-    public synchronized Optional<Result> applyCommand( ReplicatedIdAllocationRequest request, long commandIndex )
+    public synchronized void applyCommand( ReplicatedIdAllocationRequest request, long commandIndex,
+            Consumer<Result> callback )
     {
-        if( commandIndex <= state.logIndex() )
+        if ( commandIndex > state.logIndex() )
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/id/ReplicatedIdAllocationStateMachine.java]: minor: I really like the style where you check "acceptance criteria" first in a function and just return early if you don't pass, instead of wrapping the entire code block in a nested conditional.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/id/ReplicatedIdAllocationStateMachine.java [hunk 2]
```diff
@@ -43,24 +44,22 @@ public ReplicatedIdAllocationStateMachine( StateStorage<IdAllocationState> stora
...
 
-        if ( request.idRangeStart() == firstUnallocated( idType ) )
-        {
-            state.firstUnallocated( idType, request.idRangeStart() + request.idRangeLength() );
-            return Optional.of( Result.of( true ) );
-        }
-        else
-        {
-            return Optional.of( Result.of( false ) );
+            boolean result = request.idRangeStart() == firstUnallocated( idType );
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/id/ReplicatedIdAllocationStateMachine.java]: Maybe name result as requestAccepted or requestOk.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 3]
```diff
@@ -150,6 +160,24 @@ private void submitApplyJob( long lastToApply )
         } );
     }
 
+    private void handleBatch( long index, List<DistributedOperation> distributedOperations ) throws Exception
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: index -> lastIndex

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 4]
```diff
@@ -121,26 +126,31 @@ private void submitApplyJob( long lastToApply )
             {
                 lastApplyingStorage.persistStoreData( lastToApply );
 
-                while ( cursor.index() < lastToApply && cursor.next() )
+                List<DistributedOperation> distributedOperations = new ArrayList<>( maxBatchSize );
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: We might reuse a pre-allocated list and just clear it here.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 5]
```diff
@@ -121,26 +126,31 @@ private void submitApplyJob( long lastToApply )
...
-                    if ( status.isCancelled() )
+                    else
                     {
-                        return;
+                        handleBatch( cursor.index() - 1 /* ignore the current entry */, distributedOperations );
+                        assertIndexIsValidAndIncrementLastApplied( cursor.index(), 1 );
                     }
                 }
+
+                if ( !status.isCancelled() )
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: It doesn't seem super important to check for cancelling here as well really.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java [hunk 6]
```diff
@@ -191,23 +219,30 @@ public synchronized void downloadSnapshot( AdvertisedSocketAddress source ) thro
         downloader.downloadSnapshot( source, this );
     }
 
-    private void handleOperation( long commandIndex, DistributedOperation operation ) throws IOException
+    private void handleOperation( long commandIndex, List<DistributedOperation> operations )
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: handleOperation -> handleOperations

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/server/CoreEdgeClusterSettings.java [hunk 7]
```diff
@@ -183,6 +183,10 @@ public String toString()
     public static final Setting<Integer> state_machine_flush_window_size =
             setting( "core_edge.state_machine_flush_window_size", INTEGER, "100" );
 
+    @Description("The maximum number of operations to be bathed during applications of operations in the state machines")
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/server/CoreEdgeClusterSettings.java]: batched

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreStateMachines.java [hunk 8]
```diff
@@ -154,4 +142,55 @@ public void refresh( TransactionRepresentationCommitProcess localCommit )
 
         coreState.setStateMachine( this, lastAppliedIndex );
     }
+
+    private class StateMachineRequestDispatcher implements RequestDispatcher
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreStateMachines.java]: Would be good to try to explain in a comment what is being achieved here. I also wonder if replicatedTxStateMachine really is a regular simple state machine still or if it should be abstracted away under say a BatchedStateMachine abstraction which has the operations queue/sync instead or something like that.

File: enterprise/core-edge/src/test/java/org/neo4j/coreedge/raft/replication/DirectReplicator.java [hunk 9]
```diff
@@ -37,19 +37,23 @@ public DirectReplicator( StateMachine<Command> stateMachine )
     }
 
     @Override
-    public synchronized Future<Object> replicate( ReplicatedContent content, boolean trackResult )
+    public synchronized Future<Object> replicate( Command content, boolean trackResult )
     {
-        Optional<Result> result = stateMachine.applyCommand( (Command) content, commandIndex++ );
+        AtomicBoolean called = new AtomicBoolean();
```
Comment (martinfurmanski) [inline enterprise/core-edge/src/test/java/org/neo4j/coreedge/raft/replication/DirectReplicator.java]: The assertion and the variable really don't seem valuable anymore.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/id/ReplicatedIdAllocationStateMachine.java (hunk 1 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/id/ReplicatedIdAllocationStateMachine.java]: I don't mind to change it. There are contrasting opinions on this kind of style.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java (hunk 4 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreState.java]: Yeah, I forgot to move it out.

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreStateMachines.java (hunk 8 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreStateMachines.java]: Sure. Can we think about that and do the changes in a different PR?

Comment (martinfurmanski) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/state/CoreStateMachines.java]: Yes!

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/TransactionToApply.java [hunk 10]
```diff
@@ -123,11 +126,20 @@ public void commitment( Commitment commitment, long transactionId )
...
+        }
     }
 
     @Override
     public TransactionToApply next()
     {
         return nextTransactionInBatch;
     }
+
+    public void onClose( Consumer<Long> callback )
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/TransactionToApply.java]: I don't think the correct term is used here. It's called `onClose`, but looking at where it's called (from `committment(...)`) it's more of a `onCommitted`, i.e. before being applied. Where in the life cycle of a transaction do you intend to hook this in?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/TransactionToApply.java [hunk 11]
```diff
@@ -66,6 +67,8 @@
     // These fields are provided by commit process/storage engine
     private Commitment commitment;
 
+    private Consumer<Long> callback;
```
Comment (tinwelint) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/TransactionToApply.java]: Can we use `LongConsumer` instead?

**Comment by tinwelint** [comment]
@davidegrohmann Hi, I wrote two highly addressable concerns in `TransactionToApply`, please review

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/tx/ReplicatedTransactionStateMachine.java [hunk 12]
```diff
@@ -84,21 +88,41 @@ public synchronized void flush() throws IOException
...
+            callback.accept( Result.of( new TransactionFailureException( LockSessionExpired,
+                    "The lock session in the cluster has changed: [current lock session id:%d, tx lock session id:%d]",
                     currentTokenId, txLockSessionId ) ) );
         }
+        else
+        {
+            try
+            {
+                TransactionToApply transaction = new TransactionToApply( tx );
+                transaction.onClose( txId -> callback.accept( Result.of( txId ) ) );
```
Comment (lutovich) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/tx/ReplicatedTransactionStateMachine.java:103]: Can we maybe pass this callback in constructor instead?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/TransactionToApply.java (hunk 10 above)
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/TransactionToApply.java]: Maybe also name the `callback` respectively `txCommittedCallback`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/TransactionCommitment.java [hunk 13]
```diff
@@ -27,6 +29,7 @@
     private final LogPosition logPosition;
     private final TransactionIdStore transactionIdStore;
     private boolean markedAsCommitted;
+    private Consumer<Long> callback;
```
Comment (lutovich) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/TransactionCommitment.java:32]: Probably better name would be `closedCallback`

File: enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/tx/ReplicatedTransactionStateMachine.java (hunk 12 above)
Comment (davidegrohmann) [inline enterprise/core-edge/src/main/java/org/neo4j/coreedge/raft/replication/tx/ReplicatedTransactionStateMachine.java:103]: No. For the same reason you don't pass next in construction, it is not mandatory.
